spring-cloud-kubernetes-fabric8-leader/src/main/java/org/springframework/cloud/kubernetes/fabric8/leader/election/Fabric8LeaderElectionInitiator.java

```diff
@@ -0,0 +1,156 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.kubernetes.fabric8.leader.election;
+
+import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.fabric8.kubernetes.api.model.Pod;
+import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.extended.leaderelection.LeaderElectionConfig;
+import io.fabric8.kubernetes.client.extended.leaderelection.LeaderElector;
+import io.fabric8.kubernetes.client.readiness.Readiness;
+import io.fabric8.kubernetes.client.utils.CachedSingleThreadScheduler;
+import jakarta.annotation.PostConstruct;
+import jakarta.annotation.PreDestroy;
+
+import org.springframework.cloud.kubernetes.commons.leader.election.LeaderElectionProperties;
+import org.springframework.core.log.LogAccessor;
+
+/**
+ * @author wind57
+ */
+final class Fabric8LeaderElectionInitiator {
+
+	private static final LogAccessor LOG = new LogAccessor(Fabric8LeaderElectionInitiator.class);
+
+	private final CachedSingleThreadScheduler scheduler = new CachedSingleThreadScheduler();
+
+	private final String holderIdentity;
+
+	private final String podNamespace;
+
+	private final KubernetesClient fabric8KubernetesClient;
+
+	private final LeaderElectionConfig leaderElectionConfig;
+
+	private final LeaderElectionProperties leaderElectionProperties;
+
+	private final AtomicReference<ExecutorService> executorService = new AtomicReference<>();
+
+	private final AtomicReference<ScheduledFuture<?>> scheduledFuture = new AtomicReference<>();
+
+	private final AtomicReference<CompletableFuture<?>> leaderFuture = new AtomicReference<>();
+
+	Fabric8LeaderElectionInitiator(String holderIdentity, String podNamespace, KubernetesClient fabric8KubernetesClient,
+			LeaderElectionConfig leaderElectionConfig, LeaderElectionProperties leaderElectionProperties) {
+		this.holderIdentity = holderIdentity;
+		this.podNamespace = podNamespace;
+		this.fabric8KubernetesClient = fabric8KubernetesClient;
+		this.leaderElectionConfig = leaderElectionConfig;
+		this.leaderElectionProperties = leaderElectionProperties;
+	}
+
+	@PostConstruct
+	void postConstruct() {
+		LOG.info(() -> "starting leader initiator");
+		executorService.set(Executors.newSingleThreadExecutor());
+		CompletableFuture<Void> podReadyFuture = new CompletableFuture<>();
+
+		// wait until pod is ready
+		if (leaderElectionProperties.waitForPodReady()) {
```

Comment:
This seems like a useful feature.  I wonder if it makes sense for it to live in the commons package?

Honestly it would be useful to have in Fabric8 IMO